tower/src/retry/backoff.rs

```diff
@@ -0,0 +1,228 @@
+//! This module contains generic backoff utlities to be used with the retry
+//! layer.
+//!
+//! The [`Backoff`] trait is a generic way to represent backoffs that can use
+//! any timer type.
+//!
+//! [`ExponentialBackoff`] implements [`Backoff`] and provides a batteries
+//! included exponential backoff and jitter strategy.
+
+use rand::Rng;
+use rand::{rngs::SmallRng, thread_rng, SeedableRng};
+use std::fmt::Display;
+use std::future::Future;
+use std::time::Duration;
+use tokio::time;
+
+/// A backoff trait where a single mutable refernces represents a single
+/// backoff session. Implementors must also implement [`Clone`] which will
+/// reset the backoff back to the default state for the next session.
+pub trait Backoff: Clone {
+    /// The future associated with each backoff. This usually will be some sort
+    /// of timer.
+    type Future: Future<Output = ()>;
+
+    /// Initiate the next backoff in the sequence.
+    fn next_backoff(&mut self) -> Self::Future;
+}
+
+/// A jittered exponential backoff strategy.
```

Comment:
let's at least briefly define what this actually does and why you would want it? maybe something like this:
```suggestion
/// A jittered [exponential backoff] strategy.
///
/// The backoff duration will increase exponentially for every subsequent
/// backoff, up to a maximum duration. A small amount of [random jitter] is
/// added to each backoff duration, in order to avoid retry spikes.
///
/// [exponential backoff]: https://en.wikipedia.org/wiki/Exponential_backoff
/// [random jitter]: https://aws.amazon.com/blogs/architecture/exponential-backoff-and-jitter/
```